File: java/signedit/src/io/nernar/innercore/signedit/SignEdit.java
package io.nernar.innercore.signedit;

import com.zhekasmirnov.innercore.api.NativeAPI;
import com.zhekasmirnov.innercore.api.NativeTileEntity;
import com.zhekasmirnov.innercore.api.commontypes.Coords;
import com.zhekasmirnov.innercore.api.commontypes.ItemInstance;
import com.zhekasmirnov.innercore.api.log.ICLog;
import com.zhekasmirnov.innercore.api.mod.ScriptableObjectHelper;
import com.zhekasmirnov.innercore.api.mod.util.ScriptableFunctionImpl;
import com.zhekasmirnov.innercore.api.runtime.Callback;
import com.zhekasmirnov.innercore.api.unlimited.IDRegistry;
import com.zhekasmirnov.innercore.mod.build.Config;
import java.util.ArrayList;
import java.util.List;
import org.mozilla.javascript.Context;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;
import vsdum.kex.util.AddonUtils;
import vsdum.kex.natives.GlobalContext;
import vsdum.kex.natives.LocalPlayer;

/**
 * Copyright 2022-2024 Nernar (github.com/nernar)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class SignEdit {
	public static final int SIGN_TILE_ENTITY_TYPE = 4;
	public static final String[] VANILLA_SIGN_IDS = new String[] { "oak_sign", "birch_sign", "spruce_sign",
			"dark_oak_sign", "acacia_sign", "jungle_sign", "warped_sign", "crimson_sign" };

	private static final List<Integer> registeredIds = new ArrayList<>();
	private static boolean signRequiredToEdit = true;

	public static boolean isSignRequiredToEdit() {
		return signRequiredToEdit;
	}

	public static void setIsSignRequiredToEdit(boolean required) {
		signRequiredToEdit = required;
	}

	private static final class ItemUseLocal extends ScriptableFunctionImpl {
		public Object call(Context context, Scriptable scriptable, Scriptable scope, Object[] objects) {
			long playerUid = ((Long) objects[3]).longValue();
			if (NativeAPI.isSneaking(playerUid)) {
				return null;
			}

			if (signRequiredToEdit) {
				ItemInstance item = (ItemInstance) objects[1];
				Integer id = Integer.valueOf(item.getId());
				if (registeredIds.indexOf(id) == -1) {
					return null;
				}
			}

			Coords coords = (Coords) objects[0];
			int x = ((Integer) coords.get("x")).intValue();
			int y = ((Integer) coords.get("y")).intValue();
			int z = ((Integer) coords.get("z")).intValue();

			NativeTileEntity tile = NativeTileEntity.getTileEntity(x, y, z);
			if (tile != null && tile.getType() == SIGN_TILE_ENTITY_TYPE) {
				NativeAPI.preventDefault();
				LocalPlayer player = GlobalContext.getLocalPlayer();
				if (player.getPlayerPermissionLevel() > 0) {
					player.openSign(x, y, z);
				}
			}
			return null;
		}
	}

	private static final class BlocksDefined extends ScriptableFunctionImpl {
		public Object call(Context context, Scriptable scriptable, Scriptable scope, Object[] objects) {
			ScriptableObject object = ScriptableObjectHelper.createEmpty();
			object.put("isSignRequiredToEdit", object, new ScriptableFunctionImpl() {
				public Object call(Context context, Scriptable scriptable, Scriptable scope, Object[] objects) {
					return Boolean.valueOf(isSignRequiredToEdit());
				}
			});
			object.put("setIsSignRequiredToEdit", object, new ScriptableFunctionImpl() {
				public Object call(Context context, Scriptable scriptable, Scriptable scope, Object[] objects) {
					setIsSignRequiredToEdit(((Boolean) objects[0]).booleanValue());
					return null;
				}
			});
			object.put("registerSign", object, new ScriptableFunctionImpl() {
				public Object call(Context context, Scriptable scriptable, Scriptable scope, Object[] objects) {
					if (objects[0] instanceof Number) {
						registerSign(((Integer) objects[0]).intValue());
						return null;
					}
					registerSign((String) objects[0]);
					return null;
				}
			});
			object.put("isRegisteredSign", object, new ScriptableFunctionImpl() {
				public Object call(Context context, Scriptable scriptable, Scriptable scope, Object[] objects) {
					if (objects[0] instanceof Number) {
						return Boolean.valueOf(isRegisteredSign(((Integer) objects[0]).intValue()));
					}
					return Boolean.valueOf(isRegisteredSign((String) objects[0]));
				}
			});
			Callback.invokeAPICallback("API:SignEdit", new Object[] { object });
			return null;
		}
	}

	static {
		for (String vanillaId : VANILLA_SIGN_IDS) {
			registeredIds.add(AddonUtils.getNumericIdFromIdentifier(vanillaId));
		}
		Callback.addCallback("ItemUseLocal", new ItemUseLocal(), 0);
		Callback.addCallback("BlocksDefined", new BlocksDefined(), 0);
	}

	public static boolean isRegisteredSign(String namedId) {
		return isRegisteredSign(IDRegistry.getIDByName(namedId));
	}

	public static boolean isRegisteredSign(int id) {
		Integer indexId = Integer.valueOf(id);
		return registeredIds.indexOf(indexId) != -1;
	}

	public static void registerSign(int id) {
		Integer indexId = Integer.valueOf(id);
		synchronized (registeredIds) {
			if (registeredIds.indexOf(indexId) == -1) {
				registeredIds.add(indexId);
				return;
			}
		}
		ICLog.d(SignEdit.class.getSimpleName(), "Sign id " + id + " already registered");
	}

	public static void registerSign(String namedId) {
		int uid = IDRegistry.getIDByName(namedId);
		if (uid == 0) {
			ICLog.e(SignEdit.class.getSimpleName(), "Sign id " + namedId + " not even registered!",
					new NullPointerException(namedId + " == 0"));
			return;
		}
		for (String id : VANILLA_SIGN_IDS) {
			if (id.equals(namedId)) {
				ICLog.i(SignEdit.class.getSimpleName(), "Sign id " + namedId + " consist with vanilla");
				return;
			}
		}
		registerSign(uid);
	}

	public static void setConfig(Config config) {
		signRequiredToEdit = config.getBool("sign_required_to_edit");
	}
}

File: kotlin/src/io/nernar/innercore/signedit/SignEdit.kt
package io.nernar.innercore.signedit

import com.zhekasmirnov.innercore.api.*
import com.zhekasmirnov.innercore.api.commontypes.*
import com.zhekasmirnov.innercore.api.log.*
import com.zhekasmirnov.innercore.api.mod.*
import com.zhekasmirnov.innercore.api.mod.util.*
import com.zhekasmirnov.innercore.api.runtime.*
import com.zhekasmirnov.innercore.api.unlimited.*
import com.zhekasmirnov.innercore.mod.build.*
import java.lang.reflect.*
import java.util.*
import org.mozilla.javascript.*
import vsdum.kex.natives.*

class SignEdit {
	public final val SIGN_TILE_ENTITY_TYPE = 4
	public final val VANILLA_SIGN_IDS = listOf("oak_sign", "birch_sign", "spruce_sign", "dark_oak_sign",
											   "acacia_sign", "jungle_sign", "warped_sign", "crimson_sign");
	
	private val registeredIds = ArrayList<Integer>()
	private var signRequiredToEdit = true
	
	public fun isSignRequiredToEdit(): Boolean = signRequiredToEdit
	
	public fun setIsSignRequiredToEdit(required: Boolean) {
		signRequiredToEdit = required
	}
	
	final object ItemUseServer : ScriptableFunctionImpl() {
		override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
			val playerUid: Long = objects[3] as Long
			if (NativeAPI.isSneaking(playerUid)) {
				return null
			}
			
			if (SignEdit::signRequiredToEdit) {
				val item: ItemInstance = objects[1] as ItemInstance
				if (SignEdit::registeredIds.indexOf(item.getId() as Integer) == -1) {
					return null
				}
			}
			
			val coords: Coords = objects[0] as Coords
			val x: Int = coords.get("x") as Int
			val y: Int = coords.get("y") as Int
			val z: Int = coords.get("z") as Int
			
			val tile = NativeTileEntity.getTileEntity(x, y, z)
			if (tile != null && tile.getType() == SignEdit::SIGN_TILE_ENTITY_TYPE) {
				NativeAPI.preventDefault()
				try {
					val player = GlobalContext.getLocalPlayer()
					if (player.getPlayerPermissionLevel() > 0) {
						player.openSign(x, y, z)
					}
				} catch (e: Exception) {
					ICLog.e(SignEdit::class.qualifiedName, "Something went wrong when sign text replacement dialog creation", e)
				}
			}
			return null
		}
	}
	
	final object BlocksDefined : ScriptableFunctionImpl() {
		override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
			val script = ScriptableObjectHelper.createEmpty()
			script.put("isSignRequiredToEdit", script, object : ScriptableFunctionImpl() {
				override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
					return SignEdit::isSignRequiredToEdit()
				}
			});
			script.put("setIsSignRequiredToEdit", script, object : ScriptableFunctionImpl() {
				override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
					SignEdit::setIsSignRequiredToEdit(objects[0] as Boolean)
					return null
				}
			});
			script.put("registerSign", script, object : ScriptableFunctionImpl() {
				override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
					if (objects[0] is Number) {
						SignEdit::registerSign(objects[0] as Int)
						return null
					}
					SignEdit::registerSign(objects[0] as String)
					return null
				}
			});
			script.put("isRegisteredSign", script, object : ScriptableFunctionImpl() {
				override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
					if (objects[0] is Number) {
						return SignEdit::isRegisteredSign(objects[0] as Int)
					}
					return SignEdit::isRegisteredSign(objects[0] as String)
				}
			});
			Callback.invokeAPICallback("API:SignEdit", listOf(script))
			return null
		}
	}
	
	init {
		Callback.addCallback("API:KernelExtension", object : ScriptableFunctionImpl() {
			override fun call(context: Context, scriptable: Scriptable, scope: Scriptable, objects: Array<Any>): Object? {
				Callback.addCallback("ItemUseServer", ItemUseServer, 0)
				return null
			}
		}, 0);
		Callback.addCallback("BlocksDefined", BlocksDefined, 0)
		ICLog.d("Kotlin", "Welcomed you")
	}
	
	public fun isRegisteredSign(namedId: String): Boolean =
		isRegisteredSign(IDRegistry.getIDByName(namedId))
	
	public fun isRegisteredSign(indexId: Int): Boolean =
		registeredIds.indexOf(indexId as Integer) != -1
	
	public fun registerSign(indexId: Int) {
		synchronized (registeredIds) {
			val integerIndexId: Integer = indexId as Integer
			if (registeredIds.indexOf(integerIndexId) == -1) {
				registeredIds.add(integerIndexId)
				return
			}
		}
		ICLog.d(SignEdit::class.qualifiedName, "Sign id $indexId already registered")
	}
	
	public fun registerSign(namedId: String) {
		val uid = IDRegistry.getIDByName(namedId)
		if (uid == 0) {
			ICLog.e(SignEdit::class.qualifiedName, "Sign id $namedId not even registered!", NullPointerException(namedId + " == 0"))
			return
		}
		for (id in VANILLA_SIGN_IDS) {
			if (id.equals(namedId)) {
				ICLog.i(SignEdit::class.qualifiedName, "Sign id $namedId consist with vanilla")
				return
			}
		}
		registerSign(uid)
	}
	
	public fun setConfig(config: Config) {
		signRequiredToEdit = config.getBool("sign_required_to_edit")
	}
	
	public fun boot(sources: HashMap<String, String>) {
		try {
			val field = IDRegistry::class.java.getDeclaredField("vanillaIdShortcut")
			field.setAccessible(true)
			val vanillaIdShortcut: HashMap<String, Integer>? = field.get(null) as? HashMap<String, Integer>
			VANILLA_SIGN_IDS.forEach {
				val shortcut = vanillaIdShortcut?.get(it)
				registeredIds.add(shortcut as Integer)
			}
		} catch (aww: Exception) {
			throw IllegalStateException("SignEdit fail on vanilla ID registration, ensure that you are uses last version of Inner Core", aww)
		}
	}
}
